File: IUT Blagnac/2ème année/Semestre 3/ProgEff/TP1 & 2 Utilisation des dictionnaires/TP1_java_progEff_PHAKEOVILAY_Andrew_1B/tests/TestMerge.java
package tp1.tests;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import java.util.HashMap;
import java.util.Map;

import org.junit.Test;

import tp1.MapUtils;

public class TestMerge {

	@Test
	public void test_disjoint() {
		Map<String, Integer> d1 = new HashMap<>();
		d1.put("A", 1);
		d1.put("B", 2);

		Map<String, Integer> d2 = new HashMap<>();
		d2.put("C", 2);

		Map<String, Integer> expected = new HashMap<>();
		expected.put("A", 1);
		expected.put("B", 2);
		expected.put("C", 2);
		
		assertEquals(expected, MapUtils.merge(d1, d2, null));
	}
	
	@Test
	public void test_max() {
		Map<String, Integer> d1 = new HashMap<>();
		d1.put("A", 1);
		d1.put("B", 2);

		Map<String, Integer> d2 = new HashMap<>();
		d2.put("A", 3);
		d2.put("C", 2);

		Map<String, Integer> expected = new HashMap<>();
		expected.put("A", 3);
		expected.put("B", 2);
		expected.put("C", 2);
		
		assertEquals(expected, MapUtils.merge(d1, d2, Integer::max));
	}
	
	@Test
	public void test_add() {
		Map<String, Integer> d1 = new HashMap<>();
		d1.put("A", 1);
		d1.put("B", 2);

		Map<String, Integer> d2 = new HashMap<>();
		d2.put("A", 3);
		d2.put("C", 2);

		Map<String, Integer> expected = new HashMap<>();
		expected.put("A", 4);
		expected.put("B", 2);
		expected.put("C", 2);
		
		assertEquals(expected, MapUtils.merge(d1, d2, Integer::sum));
	}
	
	@Test
	public void test_conflict() {
		Map<String, Integer> d1 = new HashMap<>();
		d1.put("A", 1);
		d1.put("B", 2);

		Map<String, Integer> d2 = new HashMap<>();
		d2.put("A", 3);
		d2.put("C", 2);

		Map<String, Integer> expected = new HashMap<>();
		expected.put("A", 4);
		expected.put("B", 2);
		expected.put("C", 2);
		
		try {
			MapUtils.merge(d1, d2, null);
		} catch (IllegalArgumentException e) {
			return;
		}
		fail("Expecting IllegalArgumentException...");
	}
	
	@Test
	public void test_no_side_effect() {
		Map<String, Integer> d1 = new HashMap<>();
		d1.put("A", 1);
		d1.put("B", 2);
		Map<String, Integer> oldD1 = new HashMap<>();
		oldD1.put("A", 1);
		oldD1.put("B", 2);
		
		Map<String, Integer> d2 = new HashMap<>();
		d2.put("A", 3);
		d2.put("C", 2);
		Map<String, Integer> oldD2 = new HashMap<>();
		oldD2.put("A", 3);
		oldD2.put("C", 2);

		Map<String, Integer> expected = new HashMap<>();
		expected.put("A", 3);
		expected.put("B", 2);
		expected.put("C", 2);
		
		MapUtils.merge(d1, d2, Integer::sum);
		
		assertEquals(oldD1, d1);
		assertEquals(oldD2, d2);
		
	}


}

File: IUT Blagnac/2ème année/ProgEff/TP1 & 2 Utilisation des dictionnaires/TP2_python_progEff_PHAKEOVILAY_Andrew_1B/test_merge.py
import unittest
import dictutils

class MyTestCase(unittest.TestCase):
    
    def test_disjoint(self):
        di1 = {'A': 1, 'B': 2} 
        di2 = {'C': 2}
        res = dictutils.merge(di1, di2)
        expected = {'A': 1, 'B': 2, 'C': 2}
        self.assertEqual(expected, res)
        
    def test_max(self):
        di1 = {'A': 1, 'B': 2}
        di2 = {'A': 3, 'C': 2}
        res = dictutils.merge(di1, di2, max)
        expected = {'A': 3, 'B': 2, 'C': 2}
        self.assertEqual(expected, res)
        
    def test_add(self):
        di1 = {'A': 1, 'B': 2}
        di2 = {'A': 3, 'C': 2}
        res = dictutils.merge(di1, di2, lambda x, y: x + y)
        expected = {'A': 4, 'B': 2, 'C': 2}
        self.assertEqual(expected, res)
        
    def test_conflict(self):
        di1 = {'A': 1, 'B': 2}
        di2 = {'A': 3, 'C': 2}
        try:
            res = dictutils.merge(di1, di2)
        except ValueError:
            return  # expected error
        self.fail("Expecting ValueError, got: " + str(di))
        
    def test_no_side_effect(self):
        di1 = {'A': 1, 'B': 2}
        di2 = {'A': 3, 'C': 2}
        res = dictutils.merge(di1, di2, lambda x, y: x + y)
        expected1 = {'A': 1, 'B': 2}
        expected2 = {'A': 3, 'C': 2}
        self.assertEqual(expected1, di1)
        self.assertEqual(expected2, di2)


if __name__ == '__main__':
    unittest.main()
